tensorflow_quantum/python/optimizers/spsa_minimizer_test.py

```diff
@@ -0,0 +1,153 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Test module for tfq.python.optimizers.spsa_minimizer optimizer."""
+# Remove PYTHONPATH collisions for protobuf.
+# pylint: disable=wrong-import-position
+import sys
+NEW_PATH = [x for x in sys.path if 'com_google_protobuf' not in x]
+sys.path = NEW_PATH
+# pylint: enable=wrong-import-position
+
+from operator import mul
+from functools import reduce
+import numpy as np
+import tensorflow as tf
+from absl.testing import parameterized
+import cirq
+import sympy
+from tensorflow_quantum.python.layers.high_level import pqc
+from tensorflow_quantum.python import util
+from tensorflow_quantum.python.optimizers import spsa_minimizer
+
+
+def loss_function_with_model_parameters(model, loss, train_x, train_y):
+    """Create a new function that assign the model parameter to the model
+    and evaluate its value.
+
+    Args:
+        model : an instance of `tf.keras.Model` or its subclasses.
+        loss : a function with signature loss_value = loss(pred_y, true_y).
+        train_x : the input part of training data.
+        train_y : the output part of training data.
+
+    Returns:
+        A function that has a signature of:
+            loss_value = f(model_parameters).
+    """
+
+    # obtain the shapes of all trainable parameters in the model
+    shapes = tf.shape_n(model.trainable_variables)
+    count = 0
+    sizes = []
+
+    # Record the shape of each parameter
+    for shape in shapes:
+        n = reduce(mul, shape)
+        sizes.append(n)
+        count += n
+
+    # Function accept the parameter and evaluate model
+    @tf.function
+    def func(params):
+        """A function that can be used by tfq.optimizer.spsa_minimize.
+
+        Args:
+           params [in]: a 1D tf.Tensor.
+
+        Returns:
+            Loss function value
+        """
+
+        # update the parameters of the model
+        start = 0
+        for i, size in enumerate(sizes):
+            model.trainable_variables[i].assign(
+                tf.reshape(params[start:start + size], shape))
+            start += size
+
+        # evaluate the loss
+        loss_value = loss(model(train_x, training=True), train_y)
+        return loss_value
+
+    return func
+
+
+class SPSAMinimizerTest(tf.test.TestCase, parameterized.TestCase):
+    """Tests for the SPSA optimization algorithm."""
+
+    def test_nonlinear_function_optimization(self):
```

Comment:
Could we add three (or more) tests with simple learning functions (maybe more sinusoids?) to verify a few other key baseline functionalities:

1. If failed to converge, does the number of function iterations in the result line up with the supplied arguments
2. If failed to converge due to tolerance misses, is that in fact reflected in a case where convergence to the tolerance is very close
3. do blocking and allow_increase interact correctly with one another and produce expected results when they are enabled/disabled and changed.

Reply:
I added tests for 1, 3 but I'm not sure what you mean by 2. Like it fails to converge because it evaluates two times and the different is smaller than the tolerance?